includes/class-sensei-frontend.php

```diff
@@ -1206,69 +1206,93 @@ public function sensei_lesson_preview_title( $title = '', $id = 0 ) {
-			}
+		/**
+		 * Lets providers give their own course sign-up handler.
+		 *
+		 * @since 3.0.0
+		 *
+		 * @hook sensei_frontend_learner_enrolment_handler
+		 *
+		 * @param    {callback(int, int):bool} $handler   Frontend enrolment handler. Takes two arguments: $user_id and $course_id. Returns `true` if successful; `false` if not.
+		 * @param    {int}                     $user_id   User ID.
+		 * @param    {int}                     $course_id Course post ID.
+		 * @return   {callback(int, int):bool} Filtered handler.
+		*/
+		$learner_enrollment_handler = apply_filters(
+			'sensei_frontend_learner_enrolment_handler',
+			array( $this, 'manually_enrol_learner' ),
+			$current_user->ID,
+			$course_id
+		);
 
-			$this->data                        = new stdClass();
-			$this->data->is_user_taking_course = false;
-			if ( $student_enrolled ) {
-				$this->data->is_user_taking_course = true;
+		$student_enrolled = false;
+		if ( is_callable( $learner_enrollment_handler ) ) {
+			$student_enrolled = call_user_func( $learner_enrollment_handler, $current_user->ID, $course_id );
+		}
 
-				// Refresh page to avoid re-posting.
-				/**
-				 * Filter the URL that students are redirected to after starting a course.
-				 *
-				 * @since 1.10.0
-				 *
-				 * @hook sensei_start_course_redirect_url
-				 *
-				 * @param {string|bool}  $redirect_url URL to redirect students to after starting course. Return `false` to prevent redirect.
-				 * @param {WP_Post}      $post         Post object for course.
-				 * @return {string|bool} Filtered URL to redirect students to after starting course. Return `false` to prevent redirect.
-				 */
-				$redirect_url = apply_filters( 'sensei_start_course_redirect_url', get_permalink( $post->ID ), $post );
+		$this->data                        = new stdClass();
+		$this->data->is_user_taking_course = false;
+		if ( ! $student_enrolled ) {
+			return;
+		}
 
-				if ( 'publish' !== get_post_status( $post ) ) {
-					wp_safe_redirect( add_query_arg( 'draftcourse', 'true', $redirect_url ) );
-					exit();
-				}
-				if ( false !== $redirect_url ) {
-					?>
+		$this->data->is_user_taking_course = true;
 
-					<script type="text/javascript"> window.location = '<?php echo esc_url_raw( $redirect_url ); ?>'; </script>
+		// Refresh page to avoid re-posting.
+		/**
+		 * Filter the URL that students are redirected to after starting a course.
+		 *
+		 * @since 1.10.0
+		 *
+		 * @hook sensei_start_course_redirect_url
+		 *
+		 * @param {string|bool}  $redirect_url URL to redirect students to after starting course. Return `false` to prevent redirect.
+		 * @param {WP_Post}      $post         Post object for course.
+		 * @return {string|bool} Filtered URL to redirect students to after starting course. Return `false` to prevent redirect.
+		*/
+		$redirect_url = apply_filters( 'sensei_start_course_redirect_url', get_permalink( $post->ID ), $post );
```

Comment:
Here we need to have the original post ID to redirect the user to a proper location in the same course.
Filtered course ID might be different (in case of WPML for a translated course it is definitely so).